src/sys/freebsd.rs

```diff
@@ -1 +1,243 @@
+    pub fn extattr_set_fd(fd: c_int, attrname: c_int, attrname: *const c_char, data: *const c_void, nbytes: size_t) -> ssize_t;
+
+    pub fn extattr_list_link(path: *const c_char, attrnamespace: c_int, data: *mut c_void, nbytes: size_t) -> ssize_t;
+    pub fn extattr_get_link(path: *const c_char, attrnamespace: c_int, attrname: *const c_char, data: *mut c_void, nbytes: size_t) -> ssize_t;
+    pub fn extattr_delete_link(path: *const c_char, attrname: c_int, attrname: *const c_char) -> c_int;
+    pub fn extattr_set_link(path: *const c_char, attrname: c_int, attrname: *const c_char, data: *const c_void, nbytes: size_t) -> ssize_t;
+}
+
+pub struct XAttrs {
+    data: Box<[u8]>,
+    offset: usize,
+}
+
+impl Clone for XAttrs {
+    fn clone(&self) -> Self {
+        XAttrs {
+            data: Vec::from(&*self.data).into_boxed_slice(),
+            offset: self.offset
+        }
+    }
+
+    fn clone_from(&mut self, other: &XAttrs) {
+        self.offset = other.offset;
+
+        let mut data = mem::replace(&mut self.data, Box::new([])).into_vec();
+        data.extend(other.data.iter().cloned());
+        self.data = data.into_boxed_slice();
+    }
+}
+
+impl Iterator for XAttrs {
+    type Item = OsString;
+    fn next(&mut self) -> Option<OsString> {
+        if self.data.is_empty() {
+            return None;
+        }
+
+        if self.offset == self.data.len() {
+            return None;
+        }
+
+        let data = &self.data[self.offset..];
+        let siz = self.data[self.offset] as usize;
+
+        self.offset += siz + 1;
+        Some(OsStr::from_bytes(&data[1..siz + 1]).to_owned())
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        if self.data.len() == self.offset {
+            (0, Some(0))
+        } else {
+            (1, None)
+        }
+    }
+}
+
+fn name_to_ns(name: &OsStr) -> io::Result<(c_int, CString)> {
+    let name_str = match name.to_os_string().into_string() {
+        Ok(s) => s,
+        Err(_) => return Err(io::Error::new(io::ErrorKind::InvalidInput, "cannot convert to rust string"))
+    };
+
+    let parts: Vec<_> = name_str.split(".").collect();
+    let ns_int = match EXTATTR_NAMESPACE_NAMES.iter().position(|&s| s == parts[0]) {
+        Some(i) => i,
+        None => return Err(io::Error::new(io::ErrorKind::InvalidInput, "no matching namespace"))
+    };
+
+    Ok((ns_int as c_int, try!(name_to_c(OsStr::new(parts[1])))))
+}
+
+fn prefix_namespaces(attrs: Vec<u8>, ns: usize) -> Vec<u8> {
+    let mut siz = 0u8;
+    let mut ret = Vec::new();
+    let nsname = EXTATTR_NAMESPACE_NAMES[ns];
+
+    for item in attrs {
+        if siz == 0 {
+            siz = item;
+            ret.push(siz + 1u8 + nsname.len() as u8);    // push size of element
+            ret.extend(nsname.as_bytes());
+            ret.push('.' as u8);
+        } else {
+            ret.push(item);
+            siz -= 1;
+        }
+    }
+
+    ret
```

Comment:
There isn't really any need to do this work up front. It's probably a better idea to just store the attribute lists in three separate Vecs and prepend namespaces to attributes as queried. Otherwise we're moving a lot of bytes around unnecessarily.